bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,80 +1,302 @@
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker version is empty".to_string(),
+            });
+            return;
+        }
+
+        if !self.is_valid_version(&worker_status.version) {
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker version is invalid".to_string(),
+            });
+            return;
+        }
+
+        self.version = worker_status.version;
+
+        let worker_supported_proof_type =
+            match ProverType::try_from(worker_status.supported_proof_type) {
+                Ok(proof_type) => proof_type,
+                Err(e) => {
+                    error!(
+                        "Failed to convert worker supported proof type ({}): {}",
+                        self.address(),
+                        e
+                    );
+                    self.set_health_status(WorkerHealthStatus::Unhealthy {
+                        num_failed_attempts: failed_attempts + 1,
+                        first_fail_timestamp: Instant::now(),
+                        reason: e.to_string(),
+                    });
+                    return;
+                },
+            };
+
+        if supported_prover_type != worker_supported_proof_type {
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: format!("Unsupported prover type: {}", worker_supported_proof_type),
+            });
+            return;
+        }
+
+        self.set_health_status(WorkerHealthStatus::Healthy);
+    }
+
+    /// Sets the worker availability.
+    pub fn set_availability(&mut self, is_available: bool) {
+        self.is_available = is_available
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of retries the worker has had.
+    pub fn num_retries(&self) -> usize {
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => 0,
+            WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts,
+                first_fail_timestamp: _,
+                reason: _,
+            } => *failed_attempts,
+            WorkerHealthStatus::Unknown => 0,
+        }
+    }
+
+    /// Returns the health status of the worker.
+    pub fn health_status(&self) -> &WorkerHealthStatus {
+        &self.health_status
+    }
+
+    /// Returns the version of the worker.
+    pub fn version(&self) -> &str {
+        &self.version
+    }
+
+    /// Returns the worker availability.
```

Comment:
nit (feel free to disregard): This implies that the worker is ready to take requests, right? If so, maybe we could clarify this to avoid ambiguities between available and ready.